codec-http2/src/main/java/io/netty/handler/codec/http2/ClearTextHttp2ServerDispatcher.java

```diff
@@ -33,54 +35,49 @@
-            directBuffer(PRI_KNOWLEDGE_PEAK_LENGTH).writeBytes(PRI_KNOWLEDGE_MAGIC.getBytes(UTF_8)));
+            directBuffer(PRI_KNOWLEDGE_PEAK_LENGTH).writeBytes(PRI_KNOWLEDGE_MAGIC.array()));
+
+    private boolean pipelineConfigured;
 
     /**
-     * Peak the first n bytes of msg, and verify that if it matches to {@link #PRI_KNOWLEDGE_FIRST_BYTES}
-     * then assume client want to start HTTP/2 connection with prior knowledge,
-     * call {@link #configurePriorKnowledge(ChannelHandlerContext)} to configure pipeline
-     * for handling HTTP/2 connection. If it didn't match, then call
-     * {@link #configureUpgrade(ChannelHandlerContext)} for HTTP/2 upgrade
+     * Peak the first {@link #PRI_KNOWLEDGE_PEAK_LENGTH} bytes of msg, and verify that
+     * if it matches to {@link #PRI_KNOWLEDGE_FIRST_BYTES} then assume client want to
+     * start HTTP/2 connection with prior knowledge, call {@link #configurePriorKnowledge(ChannelHandlerContext)}
+     * to configure pipeline for handling HTTP/2 connection. If it didn't match, then
+     * call {@link #configureUpgrade(ChannelHandlerContext)} for HTTP/2 upgrade
      *
-     * @param ctx {@link ChannelHandlerContext} of this handler
-     * @param msg the message from inbound
      * @throws IllegalStateException
      *     if {@code !(msg instanceof io.netty.ByteBuf}
      *     or if {@code msg.readableBytes() < PRI_KNOWLEDGE_PEAK_LENGTH}
      */
     @Override
-    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
-        if (!(msg instanceof ByteBuf)) {
-            throw new IllegalStateException("Only can handle ByteBuf");
+    public void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
+        if (pipelineConfigured) {
+            // After pipeline configured, just forward data to next inbound handler
+            out.add(in.readBytes(in.readableBytes()));
+            return;
         }
 
-        ByteBuf inbound = (ByteBuf) msg;
-        if (inbound.readableBytes() < PRI_KNOWLEDGE_PEAK_LENGTH) {
-            throw new IllegalStateException(
-                    String.format("must have first %d bytes to determine the protocol",
-                                  PRI_KNOWLEDGE_PEAK_LENGTH));
+        if (in.readableBytes() < PRI_KNOWLEDGE_PEAK_LENGTH) {
+            return;
         }
 
         if (ByteBufUtil.equals(PRI_KNOWLEDGE_FIRST_BYTES, 0,
-                               inbound, 0, PRI_KNOWLEDGE_PEAK_LENGTH)) {
+                               in, 0, PRI_KNOWLEDGE_PEAK_LENGTH)) {
             configurePriorKnowledge(ctx);
         } else {
             configureUpgrade(ctx);
         }
 
-        ctx.fireChannelRead(msg);
-
-        // If the ctx removed before {@code ctx.fireChannelRead(msg)}, then the
-        // message will disappear because {@code ctx.next} would be the tail of
-        // pipeline.
-        ctx.pipeline().remove(this);
```

Comment:
There is a problem if we try removed the handler from the pipeline here.
When program run to this line, the inbound message hadn't pass to next handler by calling ctx.fireChannelRead.
And if we removed the handler here, then the message will never go to next handler correctly.

Is there any way to remove current handler in the pipeline, but also pass the message to next handler?
Or maybe I can try  ChannelInboundHandlerAdapter implementation, but will buffer the inbound until readable bytes enough?